x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/process/AbstractNativeProcessTests.java

```diff
@@ -82,29 +87,34 @@ public void testStart_DoNotDetectCrashWhenNoInputPipeProvided() throws Exception
         when(processPipes.getProcessInStream()).thenReturn(Optional.empty());
         try (AbstractNativeProcess process = new TestNativeProcess()) {
             process.start(executorService);
-            wait.countDown();
+        } finally {
+            mockNativeProcessLoggingStreamEnds.countDown();
+            // Not detecting a crash is confirmed in terminateExecutorService()
         }
     }
 
     public void testStart_DoNotDetectCrashWhenProcessIsBeingClosed() throws Exception {
         try (AbstractNativeProcess process = new TestNativeProcess()) {
             process.start(executorService);
-            wait.countDown();
+        } finally {
+            mockNativeProcessLoggingStreamEnds.countDown();
+            // Not detecting a crash is confirmed in terminateExecutorService()
         }
     }
 
     public void testStart_DoNotDetectCrashWhenProcessIsBeingKilled() throws Exception {
         try (AbstractNativeProcess process = new TestNativeProcess()) {
             process.start(executorService);
             process.kill();
-            wait.countDown();
+            mockNativeProcessLoggingStreamEnds.countDown();
```

Comment:
Should it be in the `finally` block just like the other tests? If not, what's the difference?

Reply:
This is testing what happens if the process is killed.  The finally block will run after `close()`.

This does actually create an inefficiency, in that if `start()` or `kill()` throws an exception then the countdown won't happen and the `@After` method will block for 10 seconds waiting for the countdown that will never come.

This was always a problem even before my changes, but I will change the test so that it doesn't use a try-with-resources so this can be avoided.

Reply:
> This does actually create an inefficiency, in that if `start()` or `kill()` throws an exception then the countdown won't happen and the `@After` method will block for 10 seconds waiting for the countdown that will never come.

Actually this isn't true.  Terminating the threadpool will interrupt all the threads in it, and that will exit the countdown `await()` with an `InterruptedException`.  So it wouldn't block for 10 seconds.

Still I will change the code as it obviously wasn't clear what was going on.